Data/ByteString/Builder/RealFloat/Internal.hs

```diff
@@ -0,0 +1,438 @@
+{-# LANGUAGE ScopedTypeVariables, ExplicitForAll #-}
+{-# LANGUAGE BangPatterns, MagicHash, UnboxedTuples #-}
+{-# LANGUAGE CPP #-}
+
+module Data.ByteString.Builder.RealFloat.Internal
+    ( (.>>)
+    , (.<<)
+    , mask
+    , asWord
+    , pmap
+    , (...)
+    , special
+    , decimalLength9
+    , decimalLength17
+    , pow5bitsUnboxed
+    , log10pow2Unboxed
+    , log10pow5Unboxed
+    , pow5_factor
+    , multipleOfPowerOf5_Unboxed
+    , multipleOfPowerOf5_UnboxedB
+    , multipleOfPowerOf2Unboxed
+    , acceptBoundsUnboxed
+    , toCharsScientific
+    -- hand-rolled division and remainder for f2s and d2s
+    , fquot10
+    , frem10
+    , fquotRem10
+    , fquotRem10Boxed
+    , fquot5
+    , frem5
+    , fquotRem5
+    , fwrapped
+    , dquot10
+    , drem10
+    , dquotRem10
+    , dquotRem10Boxed
+    , dquot5
+    , drem5
+    , dquotRem5
+    , dquot100
+    , dwrapped
+    -- prim-op helpers
+    , box
+    , unbox
+    , ByteArray(..)
+    , castDoubleToWord64
+    , castFloatToWord32
+    ) where
+
+import Control.Monad (foldM)
+import Data.Bits (Bits(..), FiniteBits(..))
+import Data.ByteString.Internal (c2w)
+import Data.ByteString.Builder.Prim.Internal (BoundedPrim, boundedPrim)
+import Data.Char (ord)
+import GHC.Int (Int32(..))
+import GHC.Exts
+import GHC.ST (ST(..), runST)
+import GHC.Word (Word8, Word32(..), Word64(..))
+import Foreign.Ptr (plusPtr)
+import qualified Foreign.Storable as S (poke)
+
+#if __GLASGOW_HASKELL__ >= 804
+import GHC.Float (castFloatToWord32, castDoubleToWord64)
+#else
+import System.IO.Unsafe (unsafePerformIO)
+import Foreign.Marshal.Utils (with)
+import Foreign.Ptr (castPtr)
+import Foreign.Storable (peek)
+{-# NOINLINE castFloatToWord32 #-}
+castFloatToWord32 :: Float -> Word32
+castFloatToWord32 x = unsafePerformIO (with x (peek . castPtr))
+
+-- | Convert a 'Double' to a 'Word64'.
+{-# NOINLINE castDoubleToWord64 #-}
+castDoubleToWord64 :: Double -> Word64
+castDoubleToWord64 x = unsafePerformIO (with x (peek . castPtr))
+#endif
+
+{-# INLINABLE (.>>) #-}
+(.>>) :: (Bits a, Integral b) => a -> b -> a
+a .>> s = unsafeShiftR a (fromIntegral s)
+
+{-# INLINABLE (.<<) #-}
+(.<<) :: (Bits a, Integral b) => a -> b -> a
+a .<< s = unsafeShiftL a (fromIntegral s)
+
+{-# INLINABLE mask #-}
+mask :: (Bits a, Integral a) => a -> a
+mask = flip (-) 1 . (.<<) 1
+
+{-# INLINABLE asWord #-}
+asWord :: Integral w => Bool -> w
+asWord = fromIntegral . fromEnum
+
+pmap :: (a -> c) -> (a, b) -> (c, b)
```

Comment:
This is just [`Control.Arrow.first`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Arrow.html#v:first)